### DISCUSSION TITLE
Nuxt server middleware to work on specific routes

### DISCUSSION THREAD

**Discussion by Ibz786** [nuxt/nuxt#23308] [Questions]
Hi all, 

Wanted to ask, is it possible in `server/middleware` when creating a particular middleware that in can be used / targeted for a specific route? i.e. Route level middleware like that in ExpressJS - https://expressjs.com/en/guide/using-middleware.html#middleware.router

For example, if we have the following setup:
```
~/server/api/
  hello/
    - index.get.ts
  random/
    - index.get.ts
 
 ~/server/middleware/
  - hello.middleware.ts
  - random.middleware.ts  
```
In `~server/api/hello/index.get.ts` I would like to run `doHelloMiddlwareFn` on `hello/index.get.ts` for example or 
in `~server/api/random/index.get.ts` I would like to run `doSomethingRandomMiddlwareFn` on `random/index.get.ts`

**Comment by gustavoconter** [comment]
Hey, I dont know if you were able to find a solution, my approach would be to create the middleware but only execute logic if the event path matches.

In your case it would be:

```typescript
export default defineEventHandler(async (event) => {
    if (!event.path.startsWith("/api/hello")) return;
    
    console.log("Hello middleware called");
});
```

**Reply by sumasagrp** [reply]
is this still the best way to achieve this ? It would have been nice to configure this like global or scoped to a path, like the page routes

**Reply by lukakazic01** [reply]
Hello, h3 added [Object Syntax](https://h3.unjs.io/guide/event-handler#object-syntax) which you can use in Nuxt. You can put your middleware in folder utils, and then just add it onRequest for that particular route or onBeforeResponse.

It would look something like this:

```
import auth from '~/server/utils/auth';

export default defineEventHandler({
  onRequest: [auth],
  handler: () => {
    return { message: 'Hello World' };
  },
});
```

Btw everything from utils folder is auto imported, i just used it for showcase.

**Reply by sumasagrp** [reply]
@lukakazic01 your example will check if the request contains a valid token / cookie for routes ? So any routes you want to to protect you would add  onRequest: [auth] ? 

If so thats really neat !

**Reply by lukakazic01** [reply]
@sumasagrp Exactly like that, whenever i need to check if user is authorized, i just add onRequest: [auth] for that route and thats it.

Im glad if it helps you !

**Reply by oom-** [reply]
@lukakazic01 
Is there a way to pass parameters to the handlers ?
Like: `onRequest: [roleCheck: ["ADMIN", "DEV"]]` ?

I suppose that this is the only way ? 
```
  onRequest: [
    (event) => apiRouteRoleCheck(event, ["ADMIN", "DEV"]),
   ],
```

Is there a way to make it trigger after the middleware under the /server/middleware folder ?

**Reply by lukakazic01** [reply]
I'm not sure if i understood correctly, but maybe[ `event.context`](https://h3.unjs.io/guide/event#eventcontext) can help you to pass parameters from one middleware to all the others in the chain, and at the end to the route you hit.
I think that global middleware's from `server/middleware` should run before any other things before hitting the route.

**Reply by oom-** [reply]
@lukakazic01 sorry I wasn't clear at all. 

So, yes you are correct, the `onRequest` hooks are called after the `middleware folder functions`.
The question was more: is there a way to put you between `middleware folder functions` and force a specific order. 

So basically find a way to manage which order and which `middleware` needs to be triggered for each route + allow to pass them custom parameters per route.

In the end, I just dropped the `middleware` folder and went this way: 
```js
export default defineEventHandler({
  onRequest: [
    // Allow only specific environnement
    (event) => blockWrongEnv(event, ["DEV", "TEST"]),
    // Retrieve redis session
    useSession,
    // Check if the role of the user is correct
    (event) => apiRouteRoleCheck(event, ["RECEPTION"]),
    // Check 2FA code
    2faCheck,
  ],
  handler: async (event) => {
     //my route
  }
});
```

The only issue with this approch is that if you forget to add a specific middleware to a route it could let it unprotected.
(So you can add a script on startup of nuxt to check that all route starting with `/server/api/admin` has at least a specific `middleware`).

I think that overall this middleware handling is not optimal and was looking for alternatives ways. So, if you have any, let me know!

**Reply by lukakazic01** [reply]
Oh nice solution there, don't have any, didn't got into this problem (yet), but will inform you if i come across!

**Reply by oom-** [reply]
@lukakazic01 ok let me know ! I came with this because of your [original answer](https://github.com/nuxt/nuxt/discussions/23308#discussioncomment-8825129), so thank you.

**Comment by arshx86** [comment]
`serverHandlers` supposed to do that but i couldnt find enough information on anywhere. Even docs are not explaining what these params does!

For example I have seperate auth middlewares, one for client `/middleware/auth.js` and one in `/server/middleware/auth.js` the client one is working well. But server middleware are also intercepting the routes that it doesn't businesses.

The auth.js in server middleware are executing for every path, not only `/api`, i find the solution with not to execute any further if `context.path` is not starting with `/api` but i don't think it is reliable.

The following code are also not working, if you have information about this snippet please help us out!

```js
  serverHandlers: [
    {
      route: "/api",
      middleware: true,
      handler: "~/server/middleware/auth.js",
    },
  ],
```

**Reply by oom-** [reply]
Hello, take a look at what I wrote just above: https://github.com/nuxt/nuxt/discussions/23308#discussioncomment-11130407

Maybe this way could help you ?

**Comment by arshx86** [comment]
It does not seems to be possible.

https://www.reddit.com/r/Nuxt/comments/1ahdq9m/comment/kooiz68/?utm_source=share&utm_medium=web3x&utm_name=web3xcss&utm_term=1&utm_content=share_button
